Add new section explaining coalesced and predicted events

patrickhlauke opened the pull request:
Closes https://github.com/w3c/pointerevents/issues/361 and https://github.com/w3c/pointerevents/issues/215


<!--
    This comment and the below content is programatically generated.
    You may add a comma-separated list of anchors you'd like a
    direct link to below (e.g. #idl-serializers, #idl-sequence):

    Don't remove this comment or modify anything below this line.
    If you don't want a preview generated for this pull request,
    just replace the whole of this comment's content by "no preview"
    and remove what's below.
-->
***
<a href="https://pr-preview.s3.amazonaws.com/w3c/pointerevents/pull/364.html" title="Last updated on May 13, 2021, 12:37 AM UTC (88ce9a0)">Preview</a> | <a href="https://pr-preview.s3.amazonaws.com/w3c/pointerevents/364/0bd62f6...88ce9a0.html" title="Last updated on May 13, 2021, 12:37 AM UTC (88ce9a0)">Diff</a>

patrickhlauke committed:
Add new explanatory section for coalesced and predicted events

- provide a simple explanation of what these things actually *are*
- shift the relevant examples into this new section and renumber the examples throughout the spec

patrickhlauke committed:
Expand "UA" to "user agent", remove one unnecessary "API"

for consistency, as "UA" appears only twice in the spec, and that use of "API" feels awkward

patrickhlauke committed:
Tweak examples intro sentence

patrickhlauke commented:
would appreciate a pair of eyes whenever your get a chance (as can't merge without one approving review) - this should be fairly non-controversial

patrickhlauke committed:
Update ECHIDNA file

mustaqahmed commented:
This is awesome!  Thanks for taking care to add this.

I have a few minor comments/questions, mostly around pulling some more text into your new section:

A. In 5.1, the big text explanation right after `getCoalescedEvents` and `getCoalescedEvents` seems to belong to the new Section 11.  Can we move the big text into it and add a link to the new section in each of `getCoalescedEvents` and `getCoalescedEvents`?

B. The image about prediction needs to show the direction of drawing, otherwise it would be very confusing.  Also, can we perhaps zoom in one part of the curve where we have two non-overlapping prediction-lines?  ("Too many ~~cooks~~points spoil the ~~broth~~image!" :smile:)

C. This is subjective, but in an introduction I prefer very focused code examples (no `mousemove`, `maxTouchPoints`, "keyboard events" etc).

patrickhlauke commented:
> A. In 5.1, the big text explanation right after getCoalescedEvents and getCoalescedEvents seems to belong to the new Section 11. Can we move the big text into it and add a link to the new section in each of getCoalescedEvents and getCoalescedEvents?

Yeah, I was considering this initially, but now wondering - should the new section here for coalesced/predicted events actually come under the (now renumbered) section 4 as a subsection, similar to 4.1.1 button states? and then yes, move the big text into it, here.

> B. The image about prediction needs to show the direction of drawing, otherwise it would be very confusing. Also, can we perhaps zoom in one part of the curve where we have two non-overlapping prediction-lines?

Agree about indication the movement direction that caused these points. Have to admit, because I was using https://patrickhlauke.github.io/touch/coalesced-predicted-events/ it ended up being surprisingly difficult to "swoosh" a good representative gesture. I might have to tweak my source script and possibly just zoom in on the end of the gesture (where the predictions become a bit clearer/better than the often quite random predictions that Chrome seems to give me during a drawing operation currently).

> C. This is subjective, but in an introduction I prefer very focused code examples (no mousemove, maxTouchPoints, "keyboard events" etc).

fair point yes, especially since we already stress the need for considering keyboard etc elsewhere. will tweak

patrickhlauke commented:
> can we perhaps zoom in one part of the curve where we have two non-overlapping prediction-lines?

checking back on this, just to mention that in my demo, i don't actually draw any prediction lines per se, just the regular lines between confirmed and non-coalesced raw points. the predictions just show as dots (as otherwise i'd need to start tweaking my drawing logic to start looking at timestamps, and start drawing line segments from the last known good point to the next predicted point only, etc). I'll see if i can make it clearer though

patrickhlauke committed:
Clean up code examples and make them minimal

patrickhlauke committed:
Merge branch 'gh-pages' into patrickhlauke-coalesced-predicted-section

patrickhlauke committed:
Indentation

patrickhlauke committed:
Add clearer picture/description for predicted points

patrickhlauke committed:
pngcrush other png images used in the spec

patrickhlauke commented:
@mustaqahmed  I chose a much cleaner example image for the predicted points and clarified in the caption what the movement that led to that line was.

patrickhlauke committed:
RMove relevant content into new section

- move the extended existing explanations for coalesced and predicted events directly into the new section
- split section into two subsections (one for coalesced events, one for predicted events)
- tweak line lengths, unnecessary use of `code` and `var`

patrickhlauke commented:
@mustaqahmed I've gone with your proposed idea of moving the content about coalesced and predicted events into the section (plus made a few tiny tweaks). Initially tried moving that whole section back up (to be a sibling section to "Button states", "Primary pointer", etc), but it felt odd / far too wordy / got in the way of getting to the actual list of "Pointer Event Types". I *think* this way here works ok.

Going to set this back from draft to ready for QA

mustaqahmed commented on index.html:
```diff
@@ -1085,6 +942,188 @@ <h3>Implicit Release of Pointer Capture</h3>
+        <section>
+            <h2>Coalesced events</h2>
+
+            <p>For performance reasons, user agents may choose not to send a move event every time
```
Perhaps better to say "...not to send a _pointermove_ event..."?

patrickhlauke replied:
Good point, yes.

mustaqahmed commented on index.html:
```diff
@@ -1085,6 +942,188 @@ <h3>Implicit Release of Pointer Capture</h3>
+            position changes into a single, averaged move event. While this approach helps in reducing
+            the amount of event handling the user agent must perform, it will naturally reduce the
+            granularity and fidelity when tracking a pointer position, particularly for fast and large
+            movements. Using the <code>getCoalescedEvents</code> method it is possible for applications
```
Link `getCoalescedEvents` back to the event interface definition?

mustaqahmed replied:
Related: some of the existing links from `getCoalescedEvents` can point to the new section 10 or one of its subsections now.  I spotted two: in `pointermove` and in `pointerrawupdate` sections.

mustaqahmed commented on index.html:
```diff
@@ -1085,6 +942,188 @@ <h3>Implicit Release of Pointer Capture</h3>
+
+            <p>A <a>PointerEvent</a> has an associated <dfn>coalesced event list</dfn> (a list of
+            zero or more <code>PointerEvents</code>). If this event is a <code>pointermove</code>
+            or <code>pointerrawupdate</code> event, it is a sequence of all <code>PointerEvent</code>
```
Nit: Replace "it" with "the list"?

patrickhlauke replied:
Good catch.

mustaqahmed commented on index.html:
```diff
@@ -1085,6 +942,188 @@ <h3>Implicit Release of Pointer Capture</h3>
+                
+                    function paint(event) {
+                        if(event.buttons>0)
+                        context.fillRect(event.clientX, event.clientY, 5, 5);
```
We need to indent this block.

patrickhlauke replied:
Done

mustaqahmed commented on index.html:
```diff
@@ -1085,6 +942,188 @@ <h3>Implicit Release of Pointer Capture</h3>
+            <p>Some user agents have built-in algorithms which, after a series of confirmed pointer movements,
+            can make a prediction (based on past points, and the speed/trajectory of the movement) what
+            the position of future pointer movements may be. Applications can use this information with
+            the <code>getPredictedEvents</code> method to speculatively "draw ahead" to a predicted position
```
Like before: link `getPredictedEvents` back to the event interface definition?

patrickhlauke replied:
I thought just doing `<a><code>getPredictedEvents</code></a>` would work, and I tried variations on this theme. based on what little i know of respec, this should work, (as there's a `<dfn>getPredictedEvents</dfn>`), but for some obtuse reason it's not and throws a respec error. Any ideas?

patrickhlauke replied:
@marcoscaceres any ideas on why just using `<a>` here isn't working?

patrickhlauke replied:
ah, never mind, counterintuitively enough, it needed specifying here somehow that the references were to this very spec here.. i.e. `<code><a data-lt="PointerEvent.getPredictedEvents">getPredictedEvents</a></code>` etc

mustaqahmed commented on index.html:
```diff
@@ -1085,6 +942,188 @@ <h3>Implicit Release of Pointer Capture</h3>
+                </code>
+            </pre>
+                
+            <p>When a <a>PointerEvent</a> is created, run the following steps for each event in the
```
The rest of Sec10.2 is related to both lists, so it seems we can start a new section here:
> 10.3 Targets of coalesced and predicted events

Then the algorithmic steps in Sec4.1 for `getCoalescedEvents` and `getPredictedEvents` entries can be moved here, to the bottom of (current 10.2) new 10.3.

After this, the definitions in 4.1 would be one liners (like "Returns the events held back by the user agent since last dispatch for performance reasons.  See section_ref" etc).

Please let me know your thoughts.  The name I proposed for the new 10.3 could be better but I have no other ideas at this moment.

patrickhlauke replied:
Made a stab at this, but not 100% convinced it all makes sense now. Made a new "Populating coalesced and predicted event lists" section, and crammed everything in there.

One aspect of the algorithmic steps that makes no sense to me, now that i look at it...what happens if the dirty flag is false? what's the target of coalesced/predicted events supposed to be then?

mustaqahmed replied:
Hmm, to me having all "list maintenance" steps in one place is way better than before.

Re what happens when the dirty bit is not set, I think it means the event goes to where it was originally meant to go.  I agree this is not clear.  But better fix that separately, it's a separate problem.  Resolving the comment.

mustaqahmed reviewed (commented):
This is very nice and organized now, thanks again!  I never realized until now how the coalesced/predicted concepts were mingled with unrelated text before!!

I have one big suggestion only, about a new 10.3 below.  The rest are minor comments.

patrickhlauke committed:
First batch of changes from mustaq's comments

patrickhlauke committed:
Add cross-reference links for getPredictedEvents and getCoalescedEvents

patrickhlauke committed:
First attempt at further reorganising getCoalescedEvents / getPredictedEvents info into new section

patrickhlauke committed:
Add more references to the new sections

mustaqahmed commented on index.html:
```diff
@@ -1085,6 +924,213 @@ <h3>Implicit Release of Pointer Capture</h3>
+                <a href="https://dom.spec.whatwg.org/#event-target"><code>target</code></a> to
+                this <code>PointerEvent</code>'s <code>target</code>.
+                <li>Set the <a>predicted events targets dirty</a> to false.
+                <li>Return the <a>predicted event list </a>
```
Nit: missing period at line end (typo from past).

patrickhlauke replied:
Good spot

mustaqahmed commented on index.html:
```diff
@@ -1085,6 +924,213 @@ <h3>Implicit Release of Pointer Capture</h3>
+        </section>
+        
+        <section>
+            <h2><dfn>Populating the coalesced and predicted event lists</dfn></h2>
```
"Maintaining"?  Not sure!

patrickhlauke replied:
While a bit awkward, I tried to combine this ... "Populating and maintaining ...". Think that's workable...

mustaqahmed reviewed (approved):
LGTM.  There are some unclear things from existing text but let's do it separately.

patrickhlauke committed:
Typo (missing full stop) and renaming new sub-section

smaug---- commented on index.html:
```diff
@@ -359,31 +301,13 @@ <h2><code>PointerEvent</code> Interface</h2>
-                                <li>Return the <a>coalesced event list</a>.</li>
-                            </ol>
-                            </p>
+                            <p>Getter method for <a>coalesced events</a>.</p>
```
Is Getter method some explicitly defined thingie? at least it isn't a getter in JS sense. Should this be "A method to access coalesced events." or some such

patrickhlauke replied:
good point, i prefer that (was only going by the previous language that was there). makes it much more understandable.

patrickhlauke replied:
changed to "A method that returns the list of ..." in both cases.

smaug---- commented on index.html:
```diff
@@ -1085,6 +924,213 @@ <h3>Implicit Release of Pointer Capture</h3>
+
+            <p>For performance reasons, user agents may choose not to send a <code>pointermove</code>
+            event every time the position of a pointer is updated. Instead, they will coalesce
+            (combine/merge) multiple position changes into a single, averaged move event. While
```
Not sure 'averaged' is what is actually happening. The event which is fired contains the most recent information about the position and such. It is just that some events which weren't fired may be put to the coalesced events list.

patrickhlauke replied:
ah, good point. i can probably just remove the ", averaged" and be done with it (as this spec doesn't define how these coalesced/non-coalesced events are divined, only how they're accessed)

patrickhlauke committed:
Suggested edits from Olli

flackr commented on index.html:
```diff
@@ -1085,6 +924,213 @@ <h3>Implicit Release of Pointer Capture</h3>
+            <p>A <a>PointerEvent</a> has an associated <dfn>coalesced events targets dirty</dfn> and an associated
+            <dfn>predicted events targets dirty</dfn> flag. When an event is created they must be initialized to false.</p>
+                
+            <p>When the <a>PointerEvent</a>'s target is changed, set <a><code>coalesced events targets dirty</code></a>
```
This seems like an implementation detail to say that the predicted and coalesced points' target must be the same as the main PointerEvent. Could we instead say that we always set the target for each coalesced /predicted event (and leave it as an implementation detail to do this lazily) or is it possible for a developer to observe when the target of these events change?

patrickhlauke replied:
Agree, this would simplify things here as well, bascially just boiling down to the statement that the target of coalesced and predicted events is always the same as its "parent" event, and that if the coalesced or predicted event was over another target, to ignore this but to set the relevant dirty flag. I'll think of a clean way to word this...

patrickhlauke replied:
> is it possible for a developer to observe when the target of these events change?

it further makes me think: is the dirty flag only set when the `getCoalescedEvents` / `getPredictedEvents` methods get called (which is what the current text suggests), or is this happening as the original / high-level / coalesced pointer event gets fired?

additionally, i've never been quite clear...where do i "find" or access this flag? is this something internal to the browser itself and not meant to be actually exposed to JS authors? this concept of the flag seems rather vague to me...

patrickhlauke replied:
(incidentally, we can either tackle this here in this PR, or make a separate issue/PR for this aspect ... whatever folks think is cleaner)

flackr replied:
Yes, as far as I can tell the flag is an internal implementation optimization, which is why I suggest we remove it from the spec and just say that we always copy the current target to the coalesced / predicted points.

I'm okay with leaving this for a new PR since it's orthogonal to the changes being made here.

patrickhlauke replied:
so purely from a JS author's point of view, what matters really is, in essence: all events in the coalesced event list and the predicted event list MUST have the same target as their "parent" pointer event (and the rest is left up to implementers on how that is achieved). maybe the whole flag business can be shuffled into a separate sub-section aimed at implementers here? (just to make the distinction clearer - at least mentioning the existence of the flag, but not going into any more detail)

patrickhlauke replied:
would also be good to come up with some name for what i call "parent" pointer event above. not sure what the generally accepted naming for this sort of thing is...

patrickhlauke replied:
@mustaqahmed @smaug---- any thoughts on these aspects? should we tackle them in a separate PR?

patrickhlauke replied:
we'll tackle this separately from https://github.com/w3c/pointerevents/issues/370

flackr commented on index.html:
```diff
@@ -1085,6 +924,213 @@ <h3>Implicit Release of Pointer Capture</h3>
+                <a href="https://dom.spec.whatwg.org/#event-target"><code>target</code></a> to
+                this <code>PointerEvent</code>'s <code>target</code>.
+                <li>Set the <a>predicted events targets dirty</a> to false.
+                <li>Return the <a>predicted event list.</a>
```
I think you have to leave the period out of the link for this link to work.

patrickhlauke replied:
well spotted. corrected.

patrickhlauke committed:
Fix broken reference link

patrickhlauke commented:
I believe that, at least in part, this PR also addresses https://github.com/w3c/pointerevents/issues/215 - please see if there's anything else that's missed out/unclear

smaug---- commented on index.html:
```diff
@@ -1085,6 +924,213 @@ <h3>Implicit Release of Pointer Capture</h3>
+            <div class="note">The PointerEvent's attributes will be initalized in a way that best represents
+            all events in the coalesced event list.
+            For example its <a href="https://www.w3.org/TR/pointerlock/#widl-MouseEvent-movementX">
+            movementX</a> and
```
For non-movement, shouldn't coordinates be the most recent ones, not "best represents
all events in the coalesced event list"

patrickhlauke replied:
i wonder if this is an implementation detail? i.e. could user agents choose to do coalescing in terms of averaging, rather than just picking the last/most recent one? do any do that? [edit: ah, i missed the "for non-movement" part]

patrickhlauke replied:
Made a tiny tweak changing "all events" to "the events". Taking a step back though, I'm actually unclear on what this note is trying to say. Right from the start, "The PointerEvent's attributes" ... which pointer event is it referring to? The coalesced "parent" pointer event that has the list of non-coalesced pointer events? If so, is this trying to define (in a non-normative note, but still) what a user agent could do? The spec already says that it isn't trying to define this. Wonder if the whole note should not be removed? I think it's worth opening a new issue for this though, as this question is essentially about existing text (rather than about the reshuffling around of that into a new section)

patrickhlauke committed:
Merge branch 'gh-pages' into patrickhlauke-coalesced-predicted-section

patrickhlauke committed:
Tiny wording tweak in note

patrickhlauke commented:
For the question relating to initialization (https://github.com/w3c/pointerevents/pull/364#discussion_r631141148 and https://github.com/w3c/pointerevents/pull/364#discussion_r631477745), I opened a separate issue https://github.com/w3c/pointerevents/issues/374

patrickhlauke committed:
Tweak example title and comment that used old "touch behavior" when meaning "direct manipulation behavior"
